ecs-agent/tmds/handlers/fault/v1/handlers/handlers_test.go

```diff
@@ -2047,3 +2053,262 @@ func TestCheckNetworkPacketLoss(t *testing.T) {
+					cmdExec.EXPECT().CombinedOutput().Times(1).Return([]byte(""), nil),
+				)
+			},
+		},
+	}
+
+	for _, tc := range tcs {
+		t.Run(tc.name, func(t *testing.T) {
+			// Mocks
+			ctrl := gomock.NewController(t)
+			defer ctrl.Finish()
+
+			agentState := mock_state.NewMockAgentState(ctrl)
+			metricsFactory := mock_metrics.NewMockEntryFactory(ctrl)
+
+			router := mux.NewRouter()
+			mockExec := mock_execwrapper.NewMockExec(ctrl)
+			handler := New(agentState, metricsFactory, mockExec)
+			networkConfigClient := netconfig.NewNetworkConfigClient()
+
+			tc.setAgentStateExpectations(agentState, networkConfigClient)
+			tc.setExecExpectations(mockExec, ctrl)
+
+			var startHandleMethod, stopHandleMethod func(http.ResponseWriter, *http.Request)
+			switch tc.faultType {
+			case types.BlackHolePortFaultType:
+				startHandleMethod = handler.StartNetworkBlackholePort()
+				stopHandleMethod = handler.StopNetworkBlackHolePort()
+			case types.LatencyFaultType:
+				startHandleMethod = handler.StartNetworkLatency()
+				stopHandleMethod = handler.StopNetworkLatency()
+			case types.PacketLossFaultType:
+				startHandleMethod = handler.StartNetworkPacketLoss()
+				stopHandleMethod = handler.StopNetworkPacketLoss()
+			default:
+				t.Error("Unrecognized network fault type")
+			}
+			router.HandleFunc(
+				NetworkFaultPath(tc.faultType, types.StartNetworkFaultPostfix),
+				startHandleMethod,
+			).Methods(http.MethodPost)
+
+			router.HandleFunc(
+				NetworkFaultPath(tc.faultType, types.StopNetworkFaultPostfix),
+				stopHandleMethod,
+			).Methods(http.MethodPost)
+
+			var requestBody io.Reader
+			reqBodyBytes, err := json.Marshal(tc.requestBody)
+			require.NoError(t, err)
+			requestBody = bytes.NewReader(reqBodyBytes)
+			startReq, err := http.NewRequest(http.MethodPost, fmt.Sprintf(startEndpoint, endpointId, tc.faultType), requestBody)
+			require.NoError(t, err)
+
+			ch1 := make(chan struct {
+				int
+				error
+			})
+
+			reqBodyBytes, err = json.Marshal(tc.requestBody)
+			require.NoError(t, err)
+			requestBody = bytes.NewReader(reqBodyBytes)
+			stopReq, err := http.NewRequest(http.MethodPost, fmt.Sprintf(stopEndpoint, endpointId, tc.faultType), requestBody)
+			require.NoError(t, err)
+
+			ch2 := make(chan struct {
+				int
+				error
+			})
+
+			// Make an asynchronous Start request first
+			go makeAsyncRequest(router, startReq, ch1)
+
+			// Waiting a bit before sending the stop request
+			time.Sleep(1 * time.Second)
+
+			// Make an asynchronous Stop request second
+			go makeAsyncRequest(router, stopReq, ch2)
+
+			// Waiting to get the status code of the start request
+			resp1 := <-ch1
+			require.NoError(t, resp1.error)
+			assert.Equal(t, http.StatusOK, resp1.int)
+
+			// Waiting to get the status code of the stop request
+			resp2 := <-ch2
+			require.NoError(t, resp2.error)
+			assert.Equal(t, http.StatusOK, resp2.int)
```

Comment:
I think what we really want to test is that the order of events is - 
1. Exec for first request is started but not finished yet 
2. Second request arrives
3. Exec for first request finishes 
4. Exec for second request starts 

Correct me if I am wrong but I think both requests being successful doesn't verify the order of events above? Should we send a message to a channel for each event above and then verify the order of events in the channel? For 1, 3, and 4 we can make use of `.Do()` function in mock expectations. For 2 we can have a wrapper on our request handler.

Reply:
>Correct me if I am wrong but I think both requests being successful doesn't verify the order of events above?

The ordering is being enforced by the `gomock.InOrder` of `setExecExpectations`. If stop requests finishes first then the tests will fail due to "unexpected call"/"missing calls to ...."

Reply:
I can also see in the logs when running the test manually that the start request gets started but hasn't finished yet and then soon after the request for stop request arrives. It also suggests that the first requests finishes and then the second requests starts

```
=== RUN   TestNetworkFaultRequestOrdering/network-blackhole-portrequest_ordering
1729715463156276769 [Info] logger=structured msg="Received new request for request type: start network-blackhole-port" request="{\"Port\":1234,\"Protocol\":\"tcp\",\"TrafficType\":\"ingress\"}" requestType="start network-blackhole-port" tmdsEndpointContainerID="endpointId"
1729715464156434911 [Info] logger=structured msg="Received new request for request type: stop network-blackhole-port" request="{\"Port\":1234,\"Protocol\":\"tcp\",\"TrafficType\":\"ingress\"}" requestType="stop network-blackhole-port" tmdsEndpointContainerID="endpointId"
1729715464156506590 [Debug] logger=structured msg="Successfully parsed fault request payload" request="{\"Port\":1234,\"Protocol\":\"tcp\",\"TrafficType\":\"ingress\"}"
1729715465156512572 [Info] logger=structured msg="Black hole port fault is not running" exitCode=1 netns="/some/path" command="nsenter --net=/some/path iptables -w 5 -C ingress-tcp-1234 -p tcp --dport 1234 -j DROP" output="iptables: Bad rule (does a matching rule exist in that chain?)." taskArn="taskArn"
1729715465156551259 [Info] logger=structured msg="Attempting to start network black hole port fault" netns="/some/path" chain="ingress-tcp-1234" taskArn="taskArn"
1729715465156565958 [Info] logger=structured msg="Successfully created new chain" taskArn="taskArn" command="nsenter --net=/some/path iptables -w 5 -N ingress-tcp-1234" output=""
1729715465156586641 [Info] logger=structured msg="Successfully added new rule to iptable chain" command="nsenter --net=/some/path iptables -w 5 -A ingress-tcp-1234 -p tcp -d 169.254.170.2 --dport 80 -j ACCEPT" output="" taskArn="taskArn"
1729715465156600368 [Info] logger=structured msg="Successfully added new rule to iptable chain" command="nsenter --net=/some/path iptables -w 5 -A ingress-tcp-1234 -p tcp -d 0.0.0.0/0 --dport 1234 -j DROP" output="" taskArn="taskArn"
1729715465156617018 [Info] logger=structured msg="Successfully inserted chain into built-in iptable" insertTable="INPUT" taskArn="taskArn" command="nsenter --net=/some/path iptables -w 5 -I INPUT -j ingress-tcp-1234" output=""
1729715465156661030 [Info] logger=structured msg="Successfully started fault" requestType="start network-blackhole-port" request="{\"Port\":1234,\"Protocol\":\"tcp\",\"TrafficType\":\"ingress\"}" response="{\"Status\":\"running\"}"
=== RUN   TestNetworkFaultRequestOrdering/network-latencyrequest_ordering
1729715465156860107 [Info] logger=structured msg="Black hole port fault has been found running" netns="/some/path" command="nsenter --net=/some/path iptables -w 5 -C ingress-tcp-1234 -p tcp --dport 1234 -j DROP" output="" taskArn="taskArn"
1729715465156870090 [Info] logger=structured msg="Attempting to stop network black hole port fault" netns="/some/path" chain="ingress-tcp-1234" taskArn="taskArn"
1729715465156880891 [Info] logger=structured msg="Successfully cleared iptable chain" command="nsenter --net=/some/path iptables -w 5 -F ingress-tcp-1234" output="" taskArn="taskArn"
1729715465156895119 [Info] logger=structured msg="Successfully deleted chain from table" command="nsenter --net=/some/path iptables -w 5 -D INPUT -j ingress-tcp-1234" output="" insertTable="INPUT" taskArn="taskArn"
1729715465156904928 [Info] logger=structured msg="Successfully deleted chain" command="nsenter --net=/some/path iptables -w 5 -X ingress-tcp-1234" output="" taskArn="taskArn"
1729715465156912144 [Info] logger=structured msg="Successfully stopped fault" requestType="stop network-blackhole-port" request="{\"Port\":1234,\"Protocol\":\"tcp\",\"TrafficType\":\"ingress\"}" response="{\"Status\":\"stopped\"}"
```

Reply:
I see. 

How about adding expected parameters to the first exec expectation for each fault? That will assert that the expectation was fulfilled by the right request. 

```
exec.EXPECT().CommandContext(gomock.Any(), gomock.Any(), gomock.Any()).Times(1).Return(cmdExec),
cmdExec.EXPECT().CombinedOutput().Times(1).Return([]byte(tcCommandEmptyOutput), nil),
```

can match exec attempts from either of the two requests.

Reply:
Thank you!

Reply:
Are we still aligned on adding expected arguments to exec expectation to ensure specific requests are being processed at the desired times? Or has this been deemed as unnecessary/out of scope?

Reply:
@danehlim They've been added. For example, https://github.com/aws/amazon-ecs-agent/pull/4409/files#diff-e1b9e4e2608bbd96428315bbe523c3258c0edfe31f91d40abd2bced989555796R2221-R2233

Reply:
Apologies, I was looking at a stale diff. Thanks for clarifying and providing the link!